### PR TITLE
[core] Add ClusterID to ClientCallManager [2/n]

### PR DESCRIPTION
**Pull request by vitsai**
This change adds a Cluster ID to GRPC client (client call, client call manager), and attaches it to the metadata of each ClientCall provided it is non-nil. 

Previous PR (GRPC server): https://github.com/ray-project/ray/pull/36517
Next PR (GCS server): https://github.com/ray-project/ray/pull/36535

Part 2 of breaking down [#35014](https://github.com/ray-project/ray/pull/35014) into more digestible parts.

## Related issue number
[#34763](https://github.com/ray-project/ray/issues/34763)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by fishbone — COMMENTED** [review]
LGTM. Please fix the compiling issues.

File: src/ray/common/id.h [hunk 1]
```diff
@@ -414,6 +414,23 @@ std::ostream &operator<<(std::ostream &os, const PlacementGroupID &id);
 // Restore the compiler alignment to default (8 bytes).
 #pragma pack(pop)
 
+struct SafeClusterID {
+  absl::Mutex m_;
+  ClusterID id GUARDED_BY(m_);
+
+  const ClusterID load() {
```
Comment (iycheng) [inline src/ray/common/id.h:421]: nit: You might want the load to be const and mark `m_` as mutable.

Comment (iycheng) [reply src/ray/common/id.h:421]: Should we also hide m_ and id? (mark them private?)

File: src/ray/rpc/client_call.h [hunk 2]
```diff
@@ -309,6 +326,9 @@ class ClientCallManager {
     }
   }
 
+  /// UUID of the cluster.
+  SafeClusterID cluster_id_;
```
Comment (iycheng) [inline src/ray/rpc/client_call.h:330]: Add comment here why we need cluster_id to be thread safe. Where is the race conditions.

**Review by fishbone — APPROVED** [review]

**Pull request merged**